.github: add workflow to build and test images, usin `docker-container` driver

fjmolinas opened the pull request:
This PR adds a .github action workflow to build docker images using buildx (buildkit). There are two cases:

- pull-request build
  - builds and push to a local registry
  - runs tests for images in local registry
- master build (or merged pull-request)
  - builds and push to dockerhub
  - runstest on dockerhub

I enabled cache for the different docker images. I had to use a different cache for each image otherwise it did not work properly. Currently, the cache never gets cleaned out so I have to update the cache on every build otherwise when it gets filled up it starts being less effective and eventually halts the worker. The cache can also be run against a local registry, cons of this are that it can't be shared among jobs, or workflows.

I've played around with this for a bit and found some advantages and disadvantages:

pros:
  - easy to leverage buildkit
  - with buildkit we could do multiarch builds
  - can use github cache to speed up builds: with cache https://github.com/fjmolinas/riotdocker/actions/runs/580652072 14min, without https://github.com/fjmolinas/riotdocker/actions/runs/579022954 ~30 min.
  - anyone can see the status of the build not only people with access to dockerhub

cons:
 - jobs can't share a registry so if we start splitting in jobs we will need to use tarballs(when not wanting to push) and that is apparently [less efficient](https://dev.to/dtinth/caching-docker-builds-in-github-actions-which-approach-is-the-fastest-a-research-18ei)
 - Splitting test into a separate job would require exporting to a tarball or similar.
 - can't split into workflow since currently having workflow depend on each other sucks (can only wait for completion not if it was successful)

Action trigger could also be against files, don't need to rebuild the base images if the only riotbuild changed for example. Although if there is a cache hit the lower images build take ~10s.

fjmolinas commented:
The following secrets will need to be added as well:

```
${{ secrets.DOCKERHUB_USERNAME }}
${{ secrets.DOCKERHUB_TOKEN }}
```

aabadie commented:
> The following secrets will need to be added as well:

I created an account "riotci" on dockerhub (with push rights) with a token but I don't have access to the projects settings (missing admin rights).

aabadie commented:
I don't quite get why the intermediate registry is required on pull-request. Why not simply build and run the tests in this case, without pushing at all ?
Is it possible to do things in 3 phases:
- Build on pull-request/push events
- Run the tests on pull-request/push events
- Push to DockerHub on push events only and if previous steps are successful
?
Otherwise, in the current state, on a push event, the image is already pushed on DockerHub before the tests are run and if something goes wrong in this last step, you end with a broken published image

miri64 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,161 @@
+    name: build riot docker images
+    runs-on: ubuntu-latest
+    env:
+      RIOT_BRANCH: 2020.10-branch
```
This already seems like having the potential to get outdated fast. Why not just use master?

fjmolinas replied:
Mimicking what is done in Travis ATM.

miri64 replied:
Ok, seems like people are bumping that version, but we maybe should think about automating this at least (also for other repos that step in releases)

kaspar030 replied:
we don't use master so this is somewhat pinned. the release needs to still build with riotdocker.

fjmolinas commented:
> I don't quite get why the intermediate registry is required on pull-request. 

Thanks for asking this is something I did not explain. With buildkit by default, the build result and intermediate cache will only remain internally in BuildKit. An [output](https://github.com/docker/buildx#-o---outputpath-typetypekeyvalue) needs to be specified to retrieve the result. What is analog to what we do know is `output=docker`, but in that case it only a single platform-build (and I wanted to keep multiplatform builds an option). So to do as you say we need either this local registry which is somehow the same as the workflow suggested but with possible multiplatform support.

But maybe I'm being too greedy and this is a further step.

>  Is it possible to do things in 3 phases:

By 3 phases you mean three jobs? Anyway, it is possible using `output=docker`.

> Otherwise, in the current state, on a push event, the image is already pushed on DockerHub before the tests are run and if something goes wrong in this last step, you end up with a broken published image

The image is only pushed on a push to master, so in theory that could not happen, the PR to be merged was already tested.

>     * Push to DockerHub on push events only and if previous steps are successful

We could split it, but pushing to docker hub doesn't make sense if the PR is not yet merged IMO.

fjmolinas commented:
> > I don't quite get why the intermediate registry is required on pull-request.
> 
> Thanks for asking this is something I did not explain. With buildkit by default, the build result and intermediate cache will only remain internally in BuildKit. An [output](https://github.com/docker/buildx#-o---outputpath-typetypekeyvalue) needs to be specified to retrieve the result. What is analog to what we do know is `output=docker`, but in that case it only a single platform-build (and I wanted to keep multiplatform builds an option). So to do as you say we need either this local registry which is somehow the same as the workflow suggested but with possible multiplatform support.
> 
> But maybe I'm being too greedy and this is a further step.
> 
> > Is it possible to do things in 3 phases:
> 
> By 3 phases you mean three jobs? Anyway, it is possible using `output=docker`.
> 
> > Otherwise, in the current state, on a push event, the image is already pushed on DockerHub before the tests are run and if something goes wrong in this last step, you end up with a broken published image
> 
> The image is only pushed on a push to master, so in theory that could not happen, the PR to be merged was already tested.
> 
> > ```
> > * Push to DockerHub on push events only and if previous steps are successful
> > ```
> 
> We could split it, but pushing to docker hub doesn't make sense if the PR is not yet merged IMO.

@aabadie I think I addressed your comments and split this into two jobs, you can see some runs here:

- master: https://github.com/fjmolinas/riotdocker/runs/2011482607?check_suite_focus=true
- pr: https://github.com/fjmolinas/riotdocker/actions/runs/613366203

fjmolinas commented:
ping @aabadie

kaspar030 commented:
I think it would be an improvement to consolidate building our docker images. Currently travis _and_ dockerhub do pr tests, and pushes to master then trigger dockerhub to update the images in their registry.
If github actions can do what we need, I think we should use it.
I'd be in favor of getting this right, then do a hard switch (probably after this release).
Maybe we can avoid having _three_ CI systems churn on this repo.

Is it possible to replicate the "hooks" Docker Hub is using to set some variables from the build?

fjmolinas commented:
> I think it would be an improvement to consolidate building our docker images. Currently travis _and_ dockerhub do pr tests, and pushes to master then trigger dockerhub to update the images in their registry.
> If github actions can do what we need, I think we should use it.
> I'd be in favor of getting this right, then do a hard switch (probably after this release).
> Maybe we can avoid having _three_ CI systems churn on this repo.
> 
> Is it possible to replicate the "hooks" Docker Hub is using to set some variables from the build?

@kaspar030 https://github.com/RIOT-OS/riotdocker/pull/136/commits/78ce17fbc9c5f536912d59cd4aadd94af73e5f90 does just that, you can test it out by pulling `fjmolinas/riotbuild:latest`

fjmolinas commented:
How do we move forward here?

fjmolinas commented:
ping!

kaspar030 commented:
> How do we move forward here?

I can give this more attention after the release, would you ping be then?

kaspar030 commented:
please rebase!

kaspar030 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,124 @@
+on:
+  push:
+    branches:
+      - master
```
can you add "trying" and "staging" here?

kaspar030 replied:
ah, but not push for master

miri64 replied:
Huh, I thought this is the idea of this action to upload images from master to Dockerhub.

miri64 commented on .github/workflows/build-test.yml:
```diff
@@ -0,0 +1,130 @@
+  push:
+    branches:
+      - trying
+      -staging
```
```suggestion
      - staging
```

fjmolinas commented:
> I think it would be an improvement to consolidate building our docker images. Currently travis _and_ dockerhub do pr tests, and pushes to master then trigger dockerhub to update the images in their registry.
> If github actions can do what we need, I think we should use it.
> I'd be in favor of getting this right, then do a hard switch (probably after this release).
> Maybe we can avoid having _three_ CI systems churn on this repo.
> 
> Is it possible to replicate the "hooks" Docker Hub is using to set some variables from the build?

Since part of the workflow only runs on master I would propose using a test remotely at first, so setting `DOCKERHUB_USERNAME` to a `riot-test` or something like that. If it all works as expected we can make a follow up PR, where the `hooks` and `travis` are removed as well as setting the `DOCKERHUB_USERNAME` to `riot`. What do you think?

fjmolinas commented:
> > I think it would be an improvement to consolidate building our docker images. Currently travis _and_ dockerhub do pr tests, and pushes to master then trigger dockerhub to update the images in their registry.
> > If github actions can do what we need, I think we should use it.
> > I'd be in favor of getting this right, then do a hard switch (probably after this release).
> > Maybe we can avoid having _three_ CI systems churn on this repo.
> > Is it possible to replicate the "hooks" Docker Hub is using to set some variables from the build?
> 
> Since part of the workflow only runs on master I would propose using a test remotely at first, so setting `DOCKERHUB_USERNAME` to a `riot-test` or something like that. If it all works as expected we can make a follow up PR, where the `hooks` and `travis` are removed as well as setting the `DOCKERHUB_USERNAME` to `riot`. What do you think?

You could also download that image and do some murdock tests.

fjmolinas commented:
bors try

fjmolinas commented:
bors try

fjmolinas commented:
bors try

fjmolinas commented:
bors try

fjmolinas commented:
bors try

miri64 commented:
Maybe it needs to be in master first for @bors to recognize?

kaspar030 commented:
> Maybe it needs to be in master first for @bors to recognize?

I think bors.toml needs an entry for the action status.

fjmolinas commented:
> > Maybe it needs to be in master first for @bors to recognize?
> 
> I think bors.toml needs an entry for the action status.

Yes I'm currently working on another issue, since @kaspar030 setup a test registry I realized when doing the multistage build it was using the pushed image.

Also the test stage failed not sure why either. So I'm working on this on my local fork.

fjmolinas commented:
> > > Maybe it needs to be in master first for @bors to recognize?
> > 
> > 
> > I think bors.toml needs an entry for the action status.
> 
> Yes I'm currently working on another issue, since @kaspar030 setup a test registry I realized when doing the multistage build it was using the pushed image.
> 
> Also the test stage failed not sure why either. So I'm working on this on my local fork.

Ajj its failing because its exhausting disk space because of the size of the pulled images....

fjmolinas commented:
I think the last push should have fixed the issues but sadly it makes the action look much more complicated than it should, or at least more complicated than if not using buildkit.

buildkit can use 3 kind of [drivers](https://github.com/docker/buildx/blob/master/docs/reference/buildx_create.md#-set-the-builder-driver-to-use---driver), for our concerns `docker` and `docker-container` are the interesting ones:

>- `docker` - Uses the builder that is built into the docker daemon. With this
  driver, the [`--load`](buildx_build.md#--load) flag is implied by default on
  `buildx build`. However, building multi-platform images or exporting cache is
  not currently supported.
> - `docker-container` - Uses a buildkit container that will be spawned via docker.
  With this driver, both building multi-platform images and exporting cache are
  supported. However, images built will not automatically appear in `docker images`
  (see [`build --load`](buildx_build.md#--load)).
  
If using `docker` driver we get the same feeling as if using regular docker builds but we lose two important things IMO: cache and multi-platform support.

But using `docker-container` forces us to go through a registry which is why I set up a local docker registry for the build test.

Sadly I then had to also split testing from building because github-runners only come with [14gb of disk space](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) which got filled up... by splitting the test into different jobs this gets solved, but.... it requires uploading and downloading the generated docker image which becomes the bottleneck when there is hot cache since our image is quite large it takes around 4min to upload. But uploading the image also has an advantage since it can be downloaded and tested... but then I'm not sure how much space we get for artifacts and the compressed image is around 2Gb.

Anyway this is still faster than going through docker hub and travis for building and testing. In any case, what I would like an opinion on from other maintainers is if you feel buildkit is not worth it with all this hassle.

fjmolinas commented:
> I think having everything built by github actions is a win, so transitioning there would be great. If buildkit makes this way more complex, maybe save that for a later day?

BTW the action succeeded now https://github.com/RIOT-OS/riotdocker/runs/2611321166?check_suite_focus=true, there is a build test failing on hifive but that is due to testing on a realease that does not match the current toolchain in the docker image.
